### PR TITLE
Fix course search filtering by term, level, and gereq

### LINKED ISSUE #6765: "Terms" filter in Course Catalog no longer works

### PR DESCRIPTION
**Pull request by drewvolz**
Closes #6765 

Fixes incorrect results when filtering by term, level, and ge requirements

- Adds helper methods for filtering by the above criteria
- Adds more performant large list props to course search results list

### CODE REVIEW

**Review by rye — APPROVED** [review]

File: source/lib/course-search/urls.ts [hunk 1]
```diff
@@ -19,13 +20,48 @@ export let timeData = (options?: Options): Promise<string[]> =>
...
+		}) as CourseType[]
+}
+
+const findMatches = (
+	findLevel: number,
+	levels: Array<number>,
+	findgereqs: Array<string> = [],
+	gereqs: Array<string>,
+) => {
+	let foundLevels = matchesLevels(findLevel, levels)
```
Comment (hawkrives) [inline source/lib/course-search/urls.ts]: I kinda want to ask that we only run the checks that someone asked for?

File: source/views/sis/course-search/results.tsx [hunk 2]
```diff
@@ -58,6 +59,55 @@ function queriesToCourses(
...
+		() => filters.find((f) => f.key === filterKey),
+		[filterKey, filters],
+	)
+}
+
+const useSelectedTerm = (filters: FilterType<CourseType>[]) => {
+	let termFilter = useSelectedFilter('term', filters)
+
+	if (termFilter) {
+		if (termFilter.enabled) {
```
Comment (hawkrives) [inline source/views/sis/course-search/results.tsx]:
```suggestion
if (termFilter?.enabled) {
```

File: source/views/sis/course-search/results.tsx [hunk 3]
```diff
@@ -58,6 +59,55 @@ function queriesToCourses(
...
+			let termFilterSpec = termFilter.spec as ListSpecType
+			return termFilterSpec.selected.map((spec) => Number(spec.title))
+		}
+	}
+
+	return []
+}
+
+const useSelecteLevel = (filters: FilterType<CourseType>[]) => {
```
Comment (hawkrives) [inline source/views/sis/course-search/results.tsx]:
```suggestion
const useSelectedLevel = (filters: FilterType<CourseType>[]) => {
```

File: source/views/sis/course-search/results.tsx [hunk 4]
```diff
@@ -58,6 +59,55 @@ function queriesToCourses(
...
+	let levelFilter = useSelectedFilter('level', filters)
+
+	if (levelFilter) {
+		if (levelFilter.enabled) {
```
Comment (hawkrives) [inline source/views/sis/course-search/results.tsx]:
```suggestion
	if (levelFilter?.enabled) {
```

File: source/views/sis/course-search/results.tsx [hunk 5]
```diff
@@ -58,6 +59,55 @@ function queriesToCourses(
...
+	}
+
+	return []
+}
+
+const useSelectedGE = (filters: FilterType<CourseType>[]) => {
+	let geFilter = useSelectedFilter('gereqs', filters)
+
+	if (geFilter) {
+		if (geFilter.enabled) {
```
Comment (hawkrives) [inline source/views/sis/course-search/results.tsx]:
```suggestion
		if (geFilter?.enabled) {
```

File: source/lib/course-search/urls.ts (hunk 1 above)
Comment (drewvolz) [reply source/lib/course-search/urls.ts]: Great catch

**Review by hawkrives — APPROVED** [review]

**Pull request merged**